### ISSUE TITLE
gRPC Proxy connection between apps gets closed after mTLS TTL passed

### ISSUE THREAD

**Issue by Syuparn** [dapr/dapr#4197]
## In what area(s)?

/area runtime

## What version of Dapr?

1.6.0

## Expected Behavior

An app can request to another app via gRPC Proxy even after their sidecars renew mTLS certifications.

## Actual Behavior

gRPC Proxy connection closes once the sidecars renew mTLS certifications.

## Steps to Reproduce the Problem

I used a sample project to reproduce this issue (using gRPC proxy with hello-world gRPC message).  https://github.com/Syuparn/grpcproxy-sample/tree/ttl-test

- prepare a kubernetes cluster
- deploy Dapr with `workloadCertTTL: 30s` and `allowedClockSkew: 1s`
    - TTL of mTLS certifications is 31s (30s + 1s)
- deploy 2 apps using gRPC proxy
    - `server`: serves helloworld gRPC service
    - `client`: invokes helloworld to `server` via gRPC proxy every 5 seconds
- `client` receives `server`'s response at first, but connection closes after 31 seconds
- if `client` pod is killed, a new pod also receives `server`'s response only during first 31 seconds

The disconnection does not occur in HTTP invoke.

**logs**

<details><summary>app container in `client`: connection closed at 07:10:38 (click to open)</summary><div>

```
$ kubectl logs client-5b465bc45-dwmvl client
2022/02/04 07:09:38 could not greet: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [::1]:50001: connect: connection refused"
2022/02/04 07:09:43 could not greet: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp [::1]:50001: connect: connection refused"
2022/02/04 07:09:48 could not greet: rpc error: code = Unavailable desc = last resolver error: produced zero addresses
2022/02/04 07:09:53 could not greet: rpc error: code = Unavailable desc = last resolver error: produced zero addresses
2022/02/04 07:09:58 could not greet: rpc error: code = Unavailable desc = last resolver error: produced zero addresses
2022/02/04 07:10:03 could not greet: rpc error: code = Unavailable desc = last resolver error: produced zero addresses
2022/02/04 07:10:08 Greeting: Hello world
2022/02/04 07:10:13 Greeting: Hello world
2022/02/04 07:10:18 Greeting: Hello world
2022/02/04 07:10:23 Greeting: Hello world
2022/02/04 07:10:28 Greeting: Hello world
2022/02/04 07:10:33 Greeting: Hello world
2022/02/04 07:10:38 could not greet: rpc error: code = Unavailable desc = last connection error: connection closed
2022/02/04 07:10:43 could not greet: rpc error: code = Unavailable desc = last connection error: connection closed
```
</div></details>

<details><summary>sidecar container in `client`: renewed certification at 07:10:34 (click to open)</summary><div>

```
$ kubectl logs client-5b465bc45-dwmvl daprd
time="2022-02-04T07:09:46.498418011Z" level=info msg="starting Dapr Runtime -- version 1.6.0 -- commit 4bb25fab444c4f1a1bf0ffd74293dbd4fdcea580" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.498445805Z" level=info msg="log level set to: debug" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.498522222Z" level=info msg="metrics server started on :9090/" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.metrics type=log ver=1.6.0
time="2022-02-04T07:09:46.502455572Z" level=debug msg="No Access control policy specified" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.acl type=log ver=1.6.0
time="2022-02-04T07:09:46.502692107Z" level=info msg="kubernetes mode configured" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.502703399Z" level=info msg="app id: client" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.502793682Z" level=info msg="mTLS enabled. creating sidecar authenticator" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.502847055Z" level=info msg="trust anchors and cert chain extracted successfully" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.security type=log ver=1.6.0
time="2022-02-04T07:09:46.502865921Z" level=info msg="authenticator created" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.503017062Z" level=info msg="Initialized name resolution to kubernetes" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.50310941Z" level=debug msg="loading component. name: kubernetes, type: secretstores.kubernetes/v1" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.503159952Z" level=info msg="loading components" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.503583703Z" level=info msg="component loaded. name: kubernetes, type: secretstores.kubernetes/v1" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.505994923Z" level=info msg="waiting for all outstanding components to be processed" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.506017978Z" level=info msg="all outstanding components processed" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.506025502Z" level=info msg="gRPC proxy enabled" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.506066942Z" level=info msg="enabled gRPC tracing middleware" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.api type=log ver=1.6.0
time="2022-02-04T07:09:46.506087722Z" level=info msg="enabled gRPC metrics middleware" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.api type=log ver=1.6.0
time="2022-02-04T07:09:46.506112489Z" level=info msg="enabled gRPC tracing middleware" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.api type=log ver=1.6.0
time="2022-02-04T07:09:46.506125665Z" level=info msg="enabled gRPC metrics middleware" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.api type=log ver=1.6.0
time="2022-02-04T07:09:46.506137137Z" level=info msg="API gRPC server is running on port 50001" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.506290752Z" level=info msg="enabled metrics http middleware" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.http type=log ver=1.6.0
time="2022-02-04T07:09:46.506309549Z" level=info msg="enabled tracing http middleware" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.http type=log ver=1.6.0
time="2022-02-04T07:09:46.506359876Z" level=info msg="enabled metrics http middleware" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.http type=log ver=1.6.0
time="2022-02-04T07:09:46.506374814Z" level=info msg="enabled tracing http middleware" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.http type=log ver=1.6.0
time="2022-02-04T07:09:46.506382168Z" level=info msg="http server is running on port 3500" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.506395193Z" level=info msg="The request body size parameter is: 4" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.506408709Z" level=info msg="enabled gRPC tracing middleware" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:09:46.506415212Z" level=info msg="enabled gRPC metrics middleware" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:09:46.506419891Z" level=info msg="sending workload csr request to sentry" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:09:46.512616111Z" level=info msg="certificate signed successfully" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:09:46.512755439Z" level=info msg="internal gRPC server is running on port 50002" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.512785947Z" level=info msg="application protocol: http. waiting on port 80.  This will block until the app is listening on that port." app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.512783388Z" level=info msg="starting workload cert expiry watcher. current cert expires on: 2022-02-04 07:10:17 +0000 UTC" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:09:46.513115482Z" level=info msg="application discovered on port 80" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.51374609Z" level=warning msg="failed to init actors: no actor state store defined" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:09:46.513773503Z" level=info msg="dapr initialized. Status: Running. Init Elapsed 11.081126ms" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime type=log ver=1.6.0
time="2022-02-04T07:10:10.512854842Z" level=info msg="renewing certificate: requesting new cert and restarting gRPC server" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:10:10.512962569Z" level=info msg="sending workload csr request to sentry" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:10:10.519876735Z" level=info msg="certificate signed successfully" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:10:34.513000047Z" level=info msg="renewing certificate: requesting new cert and restarting gRPC server" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:10:34.513457948Z" level=info msg="sending workload csr request to sentry" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:10:34.520072498Z" level=info msg="certificate signed successfully" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:10:58.513734291Z" level=info msg="renewing certificate: requesting new cert and restarting gRPC server" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:10:58.513779799Z" level=info msg="sending workload csr request to sentry" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
time="2022-02-04T07:10:58.520776714Z" level=info msg="certificate signed successfully" app_id=client instance=client-5b465bc45-dwmvl scope=dapr.runtime.grpc.internal type=log ver=1.6.0
```
</div></details>

<details><summary>app container in `server`: received no requests after 07:10:33 (click to open)</summary><div>

```
$ kubectl logs server-6768cdb95c-8gqtf server
2022/02/04 07:09:38 server listening at [::]:50051
2022/02/04 07:10:08 Received: world
2022/02/04 07:10:13 Received: world
2022/02/04 07:10:18 Received: world
2022/02/04 07:10:23 Received: world
2022/02/04 07:10:28 Received: world
2022/02/04 07:10:33 Received: world
```
</div></details>

## Release Note

RELEASE NOTE: **FIX** gRPC proxy connection closes after mTLS TTL passed

**Comment by Syuparn** [comment]
I suppose sidecars keep using the same connection for gRPC proxy even the certification is expired.

In HTTP invoke, connections are recreated if the current one is unavailable.

https://github.com/dapr/dapr/blob/ec4424851c4ec0650505ec4674d3999489d60952/pkg/messaging/direct_messaging.go#L157

implementation of `connectionCreatorFn`

https://github.com/dapr/dapr/blob/ec4424851c4ec0650505ec4674d3999489d60952/pkg/grpc/grpc.go#L85

While in gRPC proxy, connections to remote sidecars are always reused.

https://github.com/dapr/dapr/blob/ec4424851c4ec0650505ec4674d3999489d60952/pkg/messaging/grpc_proxy.go#L101

If this really causes the disconnection and is a problem, I will make a PR.

**Comment by yaron2** [comment]
> I suppose sidecars keep using the same connection for gRPC proxy even the certification is expired.
> 
> In HTTP invoke, connections are recreated if the current one is unavailable.
> 
> https://github.com/dapr/dapr/blob/ec4424851c4ec0650505ec4674d3999489d60952/pkg/messaging/direct_messaging.go#L157
> 
> implementation of `connectionCreatorFn`
> 
> https://github.com/dapr/dapr/blob/ec4424851c4ec0650505ec4674d3999489d60952/pkg/grpc/grpc.go#L85
> 
> While in gRPC proxy, connections to remote sidecars are always reused.
> 
> https://github.com/dapr/dapr/blob/ec4424851c4ec0650505ec4674d3999489d60952/pkg/messaging/grpc_proxy.go#L101
> 
> If this really causes the disconnection and is a problem, I will make a PR.

That's the cause. A PR to fix this would be great 👍

**Comment by gunniwho** [comment]
Do you know if v1.5 affected by this issue as well?

**Comment by Syuparn** [comment]
> Do you know if v1.5 affected by this issue as well?

The connection error is reproduced in v1.5.1 too.

**Comment by gunniwho** [comment]
> > Do you know if v1.5 affected by this issue as well?
> 
> The connection error is reproduced in v1.5.1 too.

Thanks!

**Comment by yaron2** [comment]
This fix is merged, will be released with 1.7.
